File: comms/ultra96/mqtt-client.py
import aiomqtt


class MQTTClient:
    def __init__(
        self,
        broker_host="127.0.0.1",
        broker_port=1883,
        client_id=None,
        keepalive=60,
    ):
        self.broker_host = broker_host
        self.broker_port = broker_port
        self.client_id = client_id
        self.keepalive = keepalive
        self._client = None
        self._task = None
        self._connected = False
        self._client = aiomqtt.Client(
            hostname=self.broker_host,
            port=self.broker_port,
            client_id=self.client_id,
            keepalive=self.keepalive,
        )

    async def connect(self):
        if self._connected:
            return

        await self._client.__aenter__()

    async def disconnect(self):
        """Disconnect from the MQTT broker."""
        if self._client:
            await self._client.__aexit__(None, None, None)
            self._client = None

    async def publish(
        self, topic: str, payload: str, qos: int = 0, retain: bool = False
    ):
        """Publish a message to a topic."""
        if not self._client:
            raise RuntimeError("Client not connected")
        await self._client.publish(topic, payload.encode(), qos=qos, retain=retain)

    async def subscribe(self, topic, callback, qos):
        """
        Subscribe to a topic and process incoming messages with the callback.

        Args:
            topic: Topic string.
            callback: Function called as callback(topic, payload).
            qos: Quality of Service level.
        """
        if not self._client:
            raise RuntimeError("Client not connected")

        async with self._client.messages() as messages:
            await self._client.subscribe(topic, qos=qos)
            async for message in messages:
                payload = message.payload.decode()
                await callback(message.topic.value, payload)


async def message_handler(topic: str, payload: str):
    print(f"[Message] {topic} => {payload}")

File: comms/middleware/internal/mqtt_client.go
package internal

import (
	"crypto/tls"
	"fmt"
	"log"
	"time"

	mqtt "github.com/eclipse/paho.mqtt.golang"
)

type MQTTClient struct {
	Client mqtt.Client
	Host   string
	Port   int
	User   string
	Pass   string
}

type MQTTTLSConfig struct {
	CACertFile     string
	ClientCertFile string
	ClientKeyFile  string
}

func SetupMQTTTLS(config MQTTTLSConfig) (*tls.Config, error) {
	caCertPool, err := LoadCACert(config.CACertFile)
	if err != nil {
		return nil, err
	}

	clientCert, err := LoadKeyPair(config.ClientCertFile, config.ClientKeyFile)
	if err != nil {
		return nil, err
	}

	return &tls.Config{
		RootCAs:      caCertPool,
		Certificates: []tls.Certificate{clientCert},
		MinVersion:   tls.VersionTLS12,
	}, nil
}

func NewMQTTClient(clientID string, host string, port int, user string, pass string, tlsConfig *tls.Config) *MQTTClient {
	opts := mqtt.NewClientOptions()
	opts.AddBroker(fmt.Sprintf("ssl://%s:%d", host, port))
	opts.SetClientID(clientID)
	if user != "" {
		opts.SetUsername(user)
		opts.SetPassword(pass)
	}

	opts.SetTLSConfig(tlsConfig)
	opts.SetKeepAlive(2 * time.Second)
	opts.SetPingTimeout(1 * time.Second)
	opts.OnConnect = func(c mqtt.Client) {
		log.Printf("Connected to MQTT broker at %s:%v", host, port)
	}
	opts.OnConnectionLost = func(c mqtt.Client, err error) {
		log.Printf("Connection lost: %v\n", err)
	}

	client := mqtt.NewClient(opts)
	if token := client.Connect(); token.Wait() && token.Error() != nil {
		log.Fatalf("Failed to connect to MQTT Broker: %v", token.Error())
	}

	return &MQTTClient{
		Client: client,
		Host:   host,
		Port:   port,
		User:   user,
		Pass:   pass,
	}
}

// Subscribe to a topic with a callback
func (m *MQTTClient) Subscribe(topic string, qos byte, callback mqtt.MessageHandler) {
	token := m.Client.Subscribe(topic, qos, callback)
	token.Wait()
	if token.Error() != nil {
		log.Fatalf("Subscribe error: %v", token.Error())
	}
	log.Printf("Subscribed to topic: %s", topic)
}

// Publish a message to a topic
func (m *MQTTClient) Publish(topic string, qos byte, retained bool, payload interface{}) {
	token := m.Client.Publish(topic, qos, retained, payload)
	token.Wait()
	if token.Error() != nil {
		log.Printf("Publish error: %v", token.Error())
	}
}
